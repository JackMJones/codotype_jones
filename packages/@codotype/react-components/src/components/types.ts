export enum Datatype {
    // Standard datatypes
    STRING = "STRING", // Postgres, MongoDB
    TEXT = "TEXT", // Postgres, MongoDB (Equiv. to String)
    INTEGER = "INTEGER", // Postgres, MongoDB
    BIGINT = "BIGINT", // Postgres
    FLOAT = "FLOAT", // Postgres
    DECIMAL = "DECIMAL", // Postgres
    NUMERIC = "NUMERIC", // Postgres
    BOOLEAN = "BOOLEAN", // Postgres, MongoDB
    DATE = "DATE", // Postgres, MongoDB
    TIME = "TIME", // Postgres
    DATETIME = "DATETIME", // Postgres
    TIMESTAMP = "TIMESTAMP", // MongoDB
    BINARY = "BINARY", // Postgres, MongoDB
    JSON = "JSON", // Postgres
    JSONB = "JSONB", // Postgres
    OBJECT = "OBJECT", // MongoDB

    // Array datatypes
    STRING_ARRAY = "STRING_ARRAY", // Postgres, MongoDB
    TEXT_ARRAY = "TEXT_ARRAY", // Postgres, MongoDB (Equiv. to String)
    INTEGER_ARRAY = "INTEGER_ARRAY", // Postgres, MongoDB
    BIGINT_ARRAY = "BIGINT_ARRAY", // Postgres
    FLOAT_ARRAY = "FLOAT_ARRAY", // Postgres
    DECIMAL_ARRAY = "DECIMAL_ARRAY", // Postgres
    NUMERIC_ARRAY = "NUMERIC_ARRAY", // Postgres
    BOOLEAN_ARRAY = "BOOLEAN_ARRAY", // Postgres, MongoDB
    DATE_ARRAY = "DATE_ARRAY", // Postgres, MongoDB
    TIME_ARRAY = "TIME_ARRAY", // Postgres
    DATETIME_ARRAY = "DATETIME_ARRAY", // Postgres
    TIMESTAMP_ARRAY = "TIMESTAMP_ARRAY", // MongoDB

    // Special Datatypes
    SINGLE_FILE = "SINGLE_FILE",
    SINGLE_IMAGE = "SINGLE_IMAGE",
}

export enum OptionType {
    STRING = "STRING",
    NUMBER = "NUMBER",
    BOOLEAN = "BOOLEAN",
    DROPDOWN = "DROPDOWN",
    MULTI_DROPDOWN = "MULTI_DROPDOWN",
    COLLECTION = "COLLECTION",
    INSTANCE = "INSTANCE",
}

export type OptionValue =
    | string
    | string[]
    | number
    | number[]
    | boolean
    | OptionValueInstance
    | OptionValueInstance[]
    | null;

export type OptionValueInstance =
    | OptionValueInstanceStandard
    | OptionValueInstanceAllowDisable;

interface OptionValueInstanceStandard {
    [key: string]: OptionValue;
}

interface OptionValueInstanceAllowDisable {
    enabled: boolean;
    value: {
        [key: string]: OptionValue;
    };
}

export interface DropdownOption {
    label: string;
    value: string;
}

export interface ConfigurationGroupProperty {
    label: string;
    identifier: string;
    description: string;
    icon: string;
    type: OptionType;
    defaultValue: OptionValue;
    required: boolean;
    enabled: boolean;
    allowDisable: boolean;
    properties: ConfigurationGroupProperty[];
    dropdownOptions: DropdownOption[];
}

export interface ConfigurationGroup {
    label: string;
    identifier: string;
    description: string;
    documentation: string; // Markdown
    enabled: boolean;
    allowDisable: boolean;
    scope: string; // enum
    variant: "LIST" | "SIDEBYSIDE"; // TODO - enum + rename!
    properties: ConfigurationGroupProperty[];
}

export interface ProjectConfiguration {
    [key: string]: OptionValueInstance;
}

export interface Attribute {
    id: string;
    label: string;
    identifier: string;
    description: string;
    required: boolean;
    unique: boolean;
    datatype: Datatype | null;
    default_value: null | string | boolean | number;
    datatypeOptions: { [key: string]: any };
    locked: boolean;
}

export interface Relation {
    id: string;
    label: string;
    identifier: string;
    description: string;
    required: boolean;
    unique: boolean;
    datatype: Datatype | null;
    default_value: null | string | boolean | number;
    datatypeOptions: { [key: string]: any };
    locked: boolean;
}

export interface Schema {
    // id: null | string;
    id: string;
    locked: boolean;
    source: string; // SchemaSource enum;
    removable: boolean;
    tokens: any; // SchemaTokenCasing;
    attributes: Attribute[];
    relations: Relation[];
    configuration: ProjectConfiguration;
    // reverse_relations: any[]; <--- these are _derived_, a computed value - doesn't need to be stored
}

export interface Project {
    label: string;
    generatorId: string;
    generatorVersion: string;
    configuration: ProjectConfiguration;
    schemas: Schema[];
}

export enum ExperienceRecommendation {
    BEGINNER = "beginner",
    JUNIOR = "junior",
    INTERMEDIATE = "intermediate",
    EXPERT = "expert",
}

export interface SchemaConfigurationGroup {
    configurationGroups: any[]; // ConfigurationGroup
    // defaultSchemas: any; // a Schemas array containing default schemas to load with the project.Learn more about default schema behavior here
    supportedDatatypes: any[]; // The datatypes supported by this generator.Only an array of DATATYPE_ * identifiers that correspond to values defined in @codotype/types are accepted.
    supportedRelations: any[]; // The relation types supported by this generator.Only an array of RELATION_TYPE_ * identifiers that correspond to values defined in @codotype/types are accepted.
}

export interface GeneratorMeta {
    id: string; // unique ID for the generator
    label: string; // short label for the generator
    description: string; // brief description of the generator
    icon: string; // URL to the generator's icon. Must be at least 200x200px
    homepage: string; // the "homepage" for this generator
    version: string; // the current version of the generator
    created_by: string; // optional (replaces "official")
    tech_tags: any; // an array of strings describing the tech used in the generator
    type_tags: any; // describes the type of codebase produced by this generator
    experience: ExperienceRecommendation; // an optional tag detailing the level of experience required to use the code produced by the generator
    project_path: string; // the name of the directory for the generator's output
    schemaConfigurationGroup: SchemaConfigurationGroup;
    configuration_groups: ConfigurationGroup[]; // an array of OptionGroup objects that expose additional configuration provided by the generator
    // All of this gets merged into configuration groups
    // defaultSchemas: any; // a Schemas array containing default schemas to load with the project.Learn more about default schema behavior here
    // defaultConfiguration: any; // object that can provide optional defaults / examples for each ConfigurationGroup.This is where you can supply default Addon data for different ConfigurationGroups
    // supportedDatatypes: any; // The datatypes supported by this generator.Only an array of DATATYPE_ * identifiers that correspond to values defined in @codotype/types are accepted.
    // supportedRelations: any; // The relation types supported by this generator.Only an array of RELATION_TYPE_ * identifiers that correspond to values defined in @codotype/types are accepted.
}
