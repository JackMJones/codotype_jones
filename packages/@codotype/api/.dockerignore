# ignore local generator repositories
registered_generators/**/*
registered_generators
