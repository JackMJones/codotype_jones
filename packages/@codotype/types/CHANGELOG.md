### v0.2.1

#### Fixes
* Typo fixed in `datatype-meta.js`
* Cleaned up formatting in `relation-meta.js`

---

### v0.2.0

#### Features
* Improved iconography and descriptions for datatypes

#### Deprecations
* Deprecated `DATATYPE_DOUBLE` type - merged functionality with `DATATYPE_FLOAT`

---

### v0.1.1

#### Features
* Added basic testing infrastructure with Mocha, Chai, & NYC

---

### v0.1.0

#### Features
* Added `@codotype/types` module
