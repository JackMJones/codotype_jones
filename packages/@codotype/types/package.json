{
  "name": "@codotype/types",
  "version": "0.1.1",
  "description": "Datatype definitions that power Codotype",
  "main": "index.js",
  "homepage": "https://github.com/codotype/codotype#readme",
  "directories": {
    "lib": "lib"
  },
  "scripts": {
    "test": "nyc mocha --exit"
  },
  "author": "Alexander Schwartzberg",
  "license": "MIT",
  "devDependencies": {
    "chai": "^4.1.2",
    "mocha": "^5.2.0",
    "nyc": "^13.0.1"
  }
}
