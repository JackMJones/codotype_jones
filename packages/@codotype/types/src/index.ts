export * from "./attribute";
export * from "./configuration-option-types";
export * from "./datatype-meta";
export * from "./datatype";
export * from "./relation-meta";
export * from "./relation";
export * from "./schema";
export * from "./ProjectConfiguration";
export * from "./project";
export * from "./generator";
export * from "./token";
export * from "./reference";
export * from "./uuid";
