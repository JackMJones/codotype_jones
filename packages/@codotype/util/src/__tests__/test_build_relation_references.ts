import {
  buildRelationReferences,
  inflateSchemasV2,
  inflateSchemaV2,
} from "../inflate";
import { Schema, RelationReference, InflatedSchema } from "@codotype/types";
import { userSchema, movieSchema } from "./test_state";

// // // //

const testCases: [string, Schema, Schema[]][] = [
  ["one word label", userSchema, [userSchema, movieSchema]],
];

describe("/lib/buildRelationReferences.js", () => {
  testCases.forEach((testCase) => {
    test(testCase[0], () => {
      const expectedReferences: RelationReference[] = buildRelationReferences({
        schema: testCase[1],
        schemas: testCase[2],
      });
      expect(expectedReferences).toMatchSnapshot();
    });
  });
});

// // // //

describe("/lib/inflateSchemaV2.js", () => {
  testCases.forEach((testCase) => {
    test(testCase[0], () => {
      const expectedSchema: InflatedSchema = inflateSchemaV2({
        schema: userSchema,
        schemas: [userSchema, movieSchema],
      });
      expect(expectedSchema).toMatchSnapshot();
    });
  });
});

// // // //

describe("/lib/inflateSchemasV2.js", () => {
  testCases.forEach((testCase) => {
    test(testCase[0], () => {
      const expectedSchemas: InflatedSchema[] = inflateSchemasV2({
        schemas: [userSchema, movieSchema],
      });
      expect(expectedSchemas).toMatchSnapshot();
    });
  });
});
