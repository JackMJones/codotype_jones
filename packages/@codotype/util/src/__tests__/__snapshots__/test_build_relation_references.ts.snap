// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/lib/buildRelationReferences.js one word label 1`] = `
Array [
  Object {
    "destinationSchemaId": "12345",
    "id": "daeda036-2332-4c12-af70-28902185f7f9",
    "identifiers": Object {
      "destination": Object {
        "alias": Object {
          "plural": Object {
            "camel": "directedMovies",
            "kebab": "directed-movies",
            "label": "Directed Movies",
            "pascal": "DirectedMovies",
            "snake": "directed_movies",
          },
          "singular": Object {
            "camel": "directedMovie",
            "kebab": "directed-movie",
            "label": "Directed Movie",
            "pascal": "DirectedMovie",
            "snake": "directed_movie",
          },
        },
        "canonical": Object {
          "plural": Object {
            "camel": "users",
            "kebab": "users",
            "label": "Users",
            "pascal": "Users",
            "snake": "users",
          },
          "singular": Object {
            "camel": "user",
            "kebab": "user",
            "label": "User",
            "pascal": "User",
            "snake": "user",
          },
        },
      },
      "source": Object {
        "alias": Object {
          "plural": Object {
            "camel": "directors",
            "kebab": "directors",
            "label": "Directors",
            "pascal": "Directors",
            "snake": "directors",
          },
          "singular": Object {
            "camel": "director",
            "kebab": "director",
            "label": "Director",
            "pascal": "Director",
            "snake": "director",
          },
        },
        "canonical": Object {
          "plural": Object {
            "camel": "movies",
            "kebab": "movies",
            "label": "Movies",
            "pascal": "Movies",
            "snake": "movies",
          },
          "singular": Object {
            "camel": "movie",
            "kebab": "movie",
            "label": "Movie",
            "pascal": "Movie",
            "snake": "movie",
          },
        },
      },
    },
    "sourceRelationId": "relation-example-01",
    "sourceSchemaId": "45678",
    "type": "TO_ONE",
  },
]
`;

exports[`/lib/inflateSchema.js one word label 1`] = `
Object {
  "attributes": Array [],
  "configuration": Object {},
  "id": "12345",
  "identifiers": Object {
    "plural": Object {
      "camel": "users",
      "kebab": "users",
      "label": "Users",
      "pascal": "Users",
      "snake": "users",
    },
    "singular": Object {
      "camel": "user",
      "kebab": "user",
      "label": "User",
      "pascal": "User",
      "snake": "user",
    },
  },
  "references": Array [
    Object {
      "destinationSchemaId": "12345",
      "id": "c8a2e20f-2dbf-47c5-87e2-9b2a004ee7e5",
      "identifiers": Object {
        "destination": Object {
          "alias": Object {
            "plural": Object {
              "camel": "directedMovies",
              "kebab": "directed-movies",
              "label": "Directed Movies",
              "pascal": "DirectedMovies",
              "snake": "directed_movies",
            },
            "singular": Object {
              "camel": "directedMovie",
              "kebab": "directed-movie",
              "label": "Directed Movie",
              "pascal": "DirectedMovie",
              "snake": "directed_movie",
            },
          },
          "canonical": Object {
            "plural": Object {
              "camel": "users",
              "kebab": "users",
              "label": "Users",
              "pascal": "Users",
              "snake": "users",
            },
            "singular": Object {
              "camel": "user",
              "kebab": "user",
              "label": "User",
              "pascal": "User",
              "snake": "user",
            },
          },
        },
        "source": Object {
          "alias": Object {
            "plural": Object {
              "camel": "directors",
              "kebab": "directors",
              "label": "Directors",
              "pascal": "Directors",
              "snake": "directors",
            },
            "singular": Object {
              "camel": "director",
              "kebab": "director",
              "label": "Director",
              "pascal": "Director",
              "snake": "director",
            },
          },
          "canonical": Object {
            "plural": Object {
              "camel": "movies",
              "kebab": "movies",
              "label": "Movies",
              "pascal": "Movies",
              "snake": "movies",
            },
            "singular": Object {
              "camel": "movie",
              "kebab": "movie",
              "label": "Movie",
              "pascal": "Movie",
              "snake": "movie",
            },
          },
        },
      },
      "sourceRelationId": "relation-example-01",
      "sourceSchemaId": "45678",
      "type": "TO_ONE",
    },
  ],
  "relations": Array [],
}
`;

exports[`/lib/inflateSchemas.js one word label 1`] = `
Array [
  Object {
    "attributes": Array [],
    "configuration": Object {},
    "id": "12345",
    "identifiers": Object {
      "plural": Object {
        "camel": "users",
        "kebab": "users",
        "label": "Users",
        "pascal": "Users",
        "snake": "users",
      },
      "singular": Object {
        "camel": "user",
        "kebab": "user",
        "label": "User",
        "pascal": "User",
        "snake": "user",
      },
    },
    "references": Array [
      Object {
        "destinationSchemaId": "12345",
        "id": "f6a58b43-be58-4172-9323-70b06e12b647",
        "identifiers": Object {
          "destination": Object {
            "alias": Object {
              "plural": Object {
                "camel": "directedMovies",
                "kebab": "directed-movies",
                "label": "Directed Movies",
                "pascal": "DirectedMovies",
                "snake": "directed_movies",
              },
              "singular": Object {
                "camel": "directedMovie",
                "kebab": "directed-movie",
                "label": "Directed Movie",
                "pascal": "DirectedMovie",
                "snake": "directed_movie",
              },
            },
            "canonical": Object {
              "plural": Object {
                "camel": "users",
                "kebab": "users",
                "label": "Users",
                "pascal": "Users",
                "snake": "users",
              },
              "singular": Object {
                "camel": "user",
                "kebab": "user",
                "label": "User",
                "pascal": "User",
                "snake": "user",
              },
            },
          },
          "source": Object {
            "alias": Object {
              "plural": Object {
                "camel": "directors",
                "kebab": "directors",
                "label": "Directors",
                "pascal": "Directors",
                "snake": "directors",
              },
              "singular": Object {
                "camel": "director",
                "kebab": "director",
                "label": "Director",
                "pascal": "Director",
                "snake": "director",
              },
            },
            "canonical": Object {
              "plural": Object {
                "camel": "movies",
                "kebab": "movies",
                "label": "Movies",
                "pascal": "Movies",
                "snake": "movies",
              },
              "singular": Object {
                "camel": "movie",
                "kebab": "movie",
                "label": "Movie",
                "pascal": "Movie",
                "snake": "movie",
              },
            },
          },
        },
        "sourceRelationId": "relation-example-01",
        "sourceSchemaId": "45678",
        "type": "TO_ONE",
      },
    ],
    "relations": Array [],
  },
  Object {
    "attributes": Array [
      Object {
        "addons": Object {
          "required": true,
        },
        "datatype": "STRING",
        "defaultValue": null,
        "id": "name-attr",
        "identifiers": Object {
          "camel": "name",
          "kebab": "name",
          "label": "Name",
          "pascal": "Name",
          "snake": "name",
        },
        "internalNote": "the name of the user",
        "locked": false,
        "source": "USER",
      },
      Object {
        "addons": Object {
          "required": true,
        },
        "datatype": "STRING",
        "defaultValue": null,
        "id": "email-attr",
        "identifiers": Object {
          "camel": "email",
          "kebab": "email",
          "label": "Email",
          "pascal": "Email",
          "snake": "email",
        },
        "internalNote": "the email of the user",
        "locked": false,
        "source": "USER",
      },
    ],
    "configuration": Object {},
    "id": "45678",
    "identifiers": Object {
      "plural": Object {
        "camel": "movies",
        "kebab": "movies",
        "label": "Movies",
        "pascal": "Movies",
        "snake": "movies",
      },
      "singular": Object {
        "camel": "movie",
        "kebab": "movie",
        "label": "Movie",
        "pascal": "Movie",
        "snake": "movie",
      },
    },
    "references": Array [],
    "relations": Array [
      Object {
        "destinationSchemaId": "12345",
        "id": "0b4bef9f-69e5-46a9-9ecd-b22238349ad9",
        "identifiers": Object {
          "destination": Object {
            "alias": Object {
              "plural": Object {
                "camel": "directors",
                "kebab": "directors",
                "label": "Directors",
                "pascal": "Directors",
                "snake": "directors",
              },
              "singular": Object {
                "camel": "director",
                "kebab": "director",
                "label": "Director",
                "pascal": "Director",
                "snake": "director",
              },
            },
            "canonical": Object {
              "plural": Object {
                "camel": "users",
                "kebab": "users",
                "label": "Users",
                "pascal": "Users",
                "snake": "users",
              },
              "singular": Object {
                "camel": "user",
                "kebab": "user",
                "label": "User",
                "pascal": "User",
                "snake": "user",
              },
            },
          },
          "source": Object {
            "alias": Object {
              "plural": Object {
                "camel": "directedMovies",
                "kebab": "directed-movies",
                "label": "Directed Movies",
                "pascal": "DirectedMovies",
                "snake": "directed_movies",
              },
              "singular": Object {
                "camel": "directedMovie",
                "kebab": "directed-movie",
                "label": "Directed Movie",
                "pascal": "DirectedMovie",
                "snake": "directed_movie",
              },
            },
            "canonical": Object {
              "plural": Object {
                "camel": "movies",
                "kebab": "movies",
                "label": "Movies",
                "pascal": "Movies",
                "snake": "movies",
              },
              "singular": Object {
                "camel": "movie",
                "kebab": "movie",
                "label": "Movie",
                "pascal": "Movie",
                "snake": "movie",
              },
            },
          },
        },
        "sourceRelationId": "relation-example-01",
        "sourceSchemaId": "45678",
        "type": "TO_ONE",
      },
    ],
  },
]
`;
